### PR TITLE
Implement WQuickXdgDecorationManagerAttached

### PR DESCRIPTION
**Pull request by justforlxz**

### CODE REVIEW

File: src/server/qtquick/private/wquickdecorationmapper_p.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+class WQuickDecorationMapper : public QQuickItem
+{
+    Q_OBJECT
+    QML_ATTACHED(WQuickDecorationMapperAttached)
+    QML_NAMED_ELEMENT(DecorationMapper)
+    QML_UNCREATABLE("Only using for attached property")
+
+public:
+    static WQuickDecorationMapperAttached *qmlAttachedProperties(QObject *target);
```
Comment (zccrs) [inline src/server/qtquick/private/wquickdecorationmapper_p.h:56]: Why add a new class not in WQuickXdgDecorationManager?

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 2]
```diff
@@ -49,20 +60,21 @@ void WQuickXdgDecorationManagerPrivate::onNewToplevelDecoration(QWXdgToplevelDec
...
-        decorat->setMode(decorat->handle()->requested_mode);
-    }
+    W_Q(WQuickXdgDecorationManager);
+
+    auto *surface = WSurface::fromHandle(decorat->handle()->toplevel->base->surface);
+    WQuickXdgDecorationManager::DecorationMode mode = static_cast<WQuickXdgDecorationManager::DecorationMode>(decorat->handle()->requested_mode);
+
+    decorations.insert(surface, mode);
+
+    Q_EMIT q->surfaceModeChanged(surface, decorations.value(surface, mode));
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:70]: Why not direct using `mode` value.

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 3]
```diff
@@ -80,20 +92,46 @@ void WQuickXdgDecorationManager::setMode(DecorationMode mode)
...
+            if (WSurface::fromHandle(wlr_decorations->toplevel->base->surface) == surface) {
+                auto * decorat = QWXdgToplevelDecorationV1::from(wlr_decorations);
+                if (mode == WQuickXdgDecorationManager::PreferClientSide)
+                    decorat->setMode(WLR_XDG_TOPLEVEL_DECORATION_V1_MODE_CLIENT_SIDE);
+                else if (mode == WQuickXdgDecorationManager::PreferServerSide)
+                    decorat->setMode(WLR_XDG_TOPLEVEL_DECORATION_V1_MODE_SERVER_SIDE);
+                else {
+                    decorat->setMode(decorat->handle()->requested_mode);
+                }
+                d->updateDecorationMode(QWXdgToplevelDecorationV1::from(wlr_decorations));
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:128]: Why not direct using `decorat`

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 4]
```diff
@@ -80,20 +92,46 @@ void WQuickXdgDecorationManager::setMode(DecorationMode mode)
...
+                break;
+            }
         }
     }
 
-    Q_EMIT modeChanged(mode);
+    d->mode = DecidesByClient;
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:134]: ?

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 5]
```diff
@@ -3,20 +3,42 @@
...
 
 WAYLIB_SERVER_BEGIN_NAMESPACE
 
+class WSurface;
+class WQuickXdgDecorationManager;
 class WQuickXdgDecorationManagerPrivate;
+class WAYLIB_SERVER_EXPORT WQuickXdgDecorationMapperAttached : public QObject
+{
+    Q_OBJECT
+    Q_PROPERTY(bool enable READ enable NOTIFY enableChanged FINAL)
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:20]: What is `enable`? `serverDecorationEnabled`?

File: src/server/qtquick/private/wquickxdgdecorationmanager_p.h [hunk 6]
```diff
@@ -3,20 +3,42 @@
...
 #include <wquickwaylandserver.h>
 
 #include <QQmlEngine>
 
 WAYLIB_SERVER_BEGIN_NAMESPACE
 
+class WSurface;
+class WQuickXdgDecorationManager;
 class WQuickXdgDecorationManagerPrivate;
+class WAYLIB_SERVER_EXPORT WQuickXdgDecorationMapperAttached : public QObject
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager_p.h:17]: `WQuickXdgDecorationManagerAttached`

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 7]
```diff
@@ -49,20 +60,21 @@ void WQuickXdgDecorationManagerPrivate::onNewToplevelDecoration(QWXdgToplevelDec
...
+
+    Q_EMIT q->surfaceModeChanged(surface, decorations.value(surface, mode));
 }
 
 WQuickXdgDecorationManager::WQuickXdgDecorationManager(QObject *parent):
     WQuickWaylandServerInterface(parent)
   , WObject(*new WQuickXdgDecorationManagerPrivate(this), nullptr)
 {
-
+    XDG_DECORATION_MANAGER = this;
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:77]:
````
if (XDG_DECORATION_MANAGER)
    qFatal(...)
````

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 8]
```diff
@@ -1,8 +1,13 @@
 // Copyright (C) 2023 rewine <luhongxu@deepin.org>.
 // SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
 
+#include "qwxdgshell.h"
```
Comment (wineee) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:4]:
```suggestion
#include <qwxdgshell.h>
```
and move behind waylib headers

File: src/server/qtquick/private/wquickxdgdecorationmanager.cpp [hunk 9]
```diff
@@ -80,20 +96,44 @@ void WQuickXdgDecorationManager::setMode(DecorationMode mode)
...
-    d->mode = mode;
+    }
 
     // update all existing decoration that mode changed
+    for (auto *surface : d->decorations.keys()) {
+        setModeBySurface(surface, mode);
+    }
+
+    d->mode = mode;
+}
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgdecorationmanager.cpp:109]: `Q_EMIT modeChanged(mode);`

**Review by zccrs — CHANGES_REQUESTED** [review]

**Review by zccrs — APPROVED** [review]

**Pull request merged**